mlrun/db/sqldb.py

```diff
@@ -0,0 +1,289 @@
+
+    @struct.setter
+    def struct(self, value):
+        self.body = pickle.dumps(value)
+
+
+# TODO: Store labels in db (one to many)
+class Artifact(Base, HasStruct):
+    __tablename__ = 'artifacts'
+
+    id = Column(Integer, primary_key=True)
+    key = Column(String)
+    project = Column(String)
+    tag = Column(String)
+    uid = Column(String)
+    updated = Column(TIMESTAMP)
+    body = Column(BLOB)
+
+
+class Function(Base, HasStruct):
+    __tablename__ = 'functions'
+
+    id = Column(Integer, primary_key=True)
+    name = Column(String)
+    project = Column(String)
+    tag = Column(String)
+    body = Column(BLOB)
+
+
+class Log(Base):
+    __tablename__ = 'logs'
+
+    id = Column(Integer, primary_key=True)
+    uid = Column(String)
+    project = Column(String)
+    body = Column(BLOB)
+
+
+class Run(Base, HasStruct):
+    __tablename__ = 'runs'
+    id = Column(Integer, primary_key=True)
+    uid = Column(String)
+    project = Column(String)
+    body = Column(BLOB)
+
+
+class SQLDB(RunDBInterface):
+    def __init__(self, dsn):
+        self.dsn = dsn
+        self.session = None
+
+    def connect(self, secrets=None):
+        engine = create_engine(self.dsn)
+        Base.metadata.create_all(engine)
+        cls = sessionmaker(bind=engine)
+        # TODO: One session per call?
+        self.session = cls()
+
+    def store_log(self, uid, project='', body=None, append=True):
+        log = self._query(Log, uid=uid, project=project).one_or_none()
+        if not log:
+            log = Log(uid=uid, project=project, body=body)
+        elif body:
+            if append:
+                log.body += body
+            else:
+                log.body = body
+        self._upsert(log)
+
+    def get_log(self, uid, project='', offset=0, size=0):
+        log = self._query(Log, uid=uid, project=project).one_or_none()
+        if not log:
+            return None
+        end = None if size == 0 else offset + size
+        return log.body[offset:end]
+
+    def store_run(self, struct, uid, project='', iter=0):
+        run = Run(
+            uid=uid,
+            project=project,
+        )
+        run.struct = struct
+        self._upsert(run)
+
+    def update_run(self, updates: dict, uid, project='', iter=0):
+        run = self._query(Run, uid=uid, project=project).one_or_none()
+        if not run:
+            raise RunDBError(f'run {uid}:{project} not found')
+        struct = run.struct
+        for key, val in updates.items():
+            update_in(struct, key, val)
+        run.struct = struct
+        self._upsert(run)
+
+    def read_run(self, uid, project='', iter=0):
+        run = self._query(Run, uid=uid, project=project).one_or_none()
+        if not run:
+            raise RunDBError(f'Run {uid}:{project} not found')
+        return run.struct
+
+    def list_runs(
+            self, name='', uid=None, project='', labels=None,
+            state='', sort=True, last=0, iter=False):
+        runs = RunList()
```

Comment:
@tebeka i assume we want to use the DB engine for doing filter (+labels filter), sort, last (limit).. also notice now there is iter=True/False in lists, e.g. limit results to iter=0 (False) or Iter=0..N (True), need to also add "since" i.e. start time after since time. one other consideration is to support paging (i.e. accept from row + limit and return total count, so the UI will be able to show page out of page for the tabels) 

this comment apply to list runs, artifacts, functions

Reply:
OK. Will do